presto-main/src/main/java/com/facebook/presto/operator/BlockEncodingBuffers.java

```diff
@@ -0,0 +1,442 @@
+        int remainingPositions = batchSize - (offset - positionsOffset);
+        int positionsToEncode = remainingPositions & ~0b111;
+        for (int i = 0; i < positionsToEncode; i += BITS_IN_BYTE) {
+            byte value = 0;
+            value |= block.isNull(positions[offset]) ? 0b1000_0000 : 0;
+            value |= block.isNull(positions[offset + 1]) ? 0b0100_0000 : 0;
+            value |= block.isNull(positions[offset + 2]) ? 0b0010_0000 : 0;
+            value |= block.isNull(positions[offset + 3]) ? 0b0001_0000 : 0;
+            value |= block.isNull(positions[offset + 4]) ? 0b0000_1000 : 0;
+            value |= block.isNull(positions[offset + 5]) ? 0b0000_0100 : 0;
+            value |= block.isNull(positions[offset + 6]) ? 0b0000_0010 : 0;
+            value |= block.isNull(positions[offset + 7]) ? 0b0000_0001 : 0;
+
+            nullsBufferContainsNull |= (value != 0);
+            nullsBufferIndex = setByte(nullsBuffer, nullsBufferIndex, value);
+            offset += BITS_IN_BYTE;
+        }
+
+        // Process the remaining positions
+        remainingPositions &= 0b111;
+        while (remainingNullsCount < remainingPositions) {
+            remainingNullsFromLastBatch[remainingNullsCount++] = block.isNull(positions[offset++]);
+        }
+    }
+
+    private void encodeNonNullsAsBits(int count)
+    {
+        if (remainingNullsCount + count < BITS_IN_BYTE) {
+            // just put all of this batch to remainingNullsFromLastBatch
+            for (int i = 0; i < count; i++) {
+                remainingNullsFromLastBatch[remainingNullsCount++] = false;
+            }
+            return;
+        }
+
+        int remainingPositions = count - encodeRemainingNullsAsBits();
+        int bytesCount = remainingPositions >>> 3;
+
+        nullsBufferIndex = fill(nullsBuffer, nullsBufferIndex, bytesCount, (byte) 0);
+
+        remainingPositions &= 0b111;
+        while (remainingNullsCount < remainingPositions) {
+            remainingNullsFromLastBatch[remainingNullsCount++] = false;
+        }
+    }
+
+    private int encodeRemainingNullsAsBits()
+    {
+        if (remainingNullsCount == 0) {
+            return 0;
+        }
+
+        byte value = 0;
+        for (int i = 0; i < remainingNullsCount; i++) {
+            value |= remainingNullsFromLastBatch[i] ? 0b1000_0000 >>> i : 0;
+        }
+
+        nullsBufferContainsNull |= (value != 0);
+        nullsBufferIndex = setByte(nullsBuffer, nullsBufferIndex, value);
+
+        int padding = BITS_IN_BYTE - remainingNullsCount;
+        remainingNullsCount = 0;
+        return padding;
+    }
+
+    protected long getNullsBufferSizeInBytes()
+    {
+        return nullsBufferIndex +                               // nulls buffer
+                remainingNullsCount;   // the remaining nulls not serialized yet
+    }
+
+    protected long getNullsBufferRetainedSizeInBytes()
+    {
+        return sizeOf(nullsBuffer) +                    // nulls buffer
```

Comment:
`// nulls buffer` and `// the remaining nulls not serialized yet` comments are redundant; they just repeat the code